### ISSUE TITLE
Transaction Condition.KeyExists(key)

### ISSUE THREAD

**Issue by camtechnet** [StackExchange/StackExchange.Redis#2450]
Hi,

I am trying to implement transaction logic (for first time) but cannot figure out how to get them to work when I am creating a key for the first time. Below is my code

var tran = _database.CreateTransaction();
tran.AddCondition(Condition.KeyExists(key));   //check if key exists
tran.KeyDeleteAsync(key);                                // if exists I want to delete
tran.HashSetAsync(key, myHashEntryArray);    // then save the data for the key
bool committed = tran.Execute();

Basically, if the key does not exist the transaction always returns false, which happens for any new key I am trying to save.

I am sure that I am missing something but can't see what. 

Any help appreciated

**Comment by mgravell** [comment]
You're adding a pre-condition that the key exists, so indeed: if it doesn't exist, I would expect it to return false. Unless I'm misunderstanding, this sound right? However, I will note: it is usually always easier to get correct atomic operations by using the Lua API - typically ScriptEvaluate[Async]; the multi/exec API (which is what the transaction approach here uses) pre-dates the Lua API, and is... tricky.

**Comment by camtechnet** [comment]
Hi, 
Thanks for clarifying. I worded the problem incorrectly. What I really wanted was guidance on how to cater for fact that key could exist or not. Looking afterwards in the unit tests I came up with this approach

        var keyExists = _database.HashValues(key).Length > 0;

        var tran = _database.CreateTransaction();
        tran.AddCondition(keyExists ? Condition.KeyExists(key) : Condition.KeyNotExists(key));

Apart from your suggestion to use LUA, is this the best way to do it?

Thanks

**Comment by mgravell** [comment]
What is the *purpose* for wanting to check that the key exists? what are you going to do differently based on that? right now, you seem to be doing the same thing either way, so... why check? What you show is one possible option, but "option for what?" is vague.

**Comment by camtechnet** [comment]
The reason for the transaction is to not proceed with the HashSetAsync if the KeyDeleteAsync fails.
I thought adding a condition was necessary, so thought best thing was to add it for the key exists /  not exists.
We are processing messages from Kafka, so I want to make sure no other message processed concurrently has modified the key.
(P.S. apologies but coming at this from RDBMS mindset, which may be the problem)

**Comment by mgravell** [comment]
It *sounds* like you don't need the constraint? To check: you want to
delete it if it exists, and then set something? You don't need a constraint
for that - just always issue the delete. If it doesn't exist, that does
nothing. So: a transaction with a delete and a hash-set, and nothing more.

Redis multi/exec indeed does not compare directly to RDBMS transactions.

**Comment by camtechnet** [comment]
OK, thanks for clarifying.
